File: src/models/transaction.py
from src.models.asset import Asset, AssetBuffer
from src.models.brokers.broker import Broker
from src.models.tax import Tax
from src.models.fee import Fee

class Transaction:
    def __init__(self, date : str, time : str, type : str, ticker : str, isin : str, shares : float, asset_currency : str, net_amount : float, net_amount_currency : str, taxes : list[Tax], fees : list[Fee], broker : Broker):
        self.date = date
        self.time = time
        self.type = type
        self.asset = Asset(ticker, isin, asset_currency)
        self.shares = shares
        self.net_amount = net_amount
        self.net_amount_currency = net_amount_currency
        self.taxes = taxes
        self.fees = fees
        self.broker = broker
        
    def fetch_data(self, asset_buffer : AssetBuffer):
        self.asset.fetch_data(asset_buffer)

    def __str__(self):
        return f"Transaction(Date: {self.date}, Time: {self.time}, Ticker: {self.asset.ticker}, ISIN: {self.asset.isin}, Shares: {self.shares}, Net Amount: {self.net_amount})"
        
    def __repr__(self):
        return f"Transaction(Date: {self.date}, Time: {self.time}, Type: {self.type}, Asset: {self.asset}, Shares: {self.shares}, Net Amount: {self.net_amount}, Net Amount Currency: {self.net_amount_currency}, Taxes: {self.taxes}, Broker: {self.broker}, Fees: {self.fees})"
    
    def to_dict(self):
        return {
            "date": self.date,
            "time": self.time,
            "type": self.type,
            "asset": self.asset,
            "shares": self.shares,
            "amount": self.net_amount,
            "amount_currency": self.net_amount_currency,
            "taxes": self.taxes,
            "broker": self.broker,
            "fees": self.fees,
        }
        
    def __eq__(self, other : 'Transaction'):
        return self.date == other.date and self.time == other.time and self.type == other.type and self.asset == other.asset and self.shares == other.shares and self.net_amount == other.net_amount and self.net_amount_currency == other.net_amount_currency and self.taxes == other.taxes and self.broker == other.broker and self.fees == other.fees
    
    def __hash__(self):
        return hash((self.date, self.time, self.type, self.asset.ticker, self.asset, self.shares, self.net_amount, self.net_amount_currency, self.broker))

File: src/models/transaction.ts
import { Asset } from "./asset.js";
import { Tax } from "./tax.js";
import { Fee } from "./fee.js";
import { Broker } from "./brokers/broker.js";
import { DateTime } from "luxon";

type TransactionType = "Buy" | "Sell" | "Dividend" | "Interest";

type Transaction = {
    id: number,
    date: DateTime;
    type: TransactionType;
    description?: string;

    asset?: Asset;        // opcional
    shares?: number;      // opcional

    amount: number; // Em caso de compra, corresponde ao valor pago pelo ativo + despesas e encargos (custo de compra final); em caso de venda, corresponde ao valor recebido pela venda do ativo - despesas e encargos (receita líquida).
    currency: string;

    broker: Broker;
    taxes?: Tax[];
    fees?: Fee[];
    exchangeRate?: number;
    matches?: Transaction[]; // campo específico para o caso de corretoras como a XTB que fornecem diretamente o match entre a compra e a venda, para evitar que sejam feitas suposições no processo de matching. Se este campo estiver preenchido, o processo de matching irá simplesmente usar as transações indicadas neste campo como correspondentes, sem fazer qualquer tipo de verificação adicional. Se este campo não estiver preenchido, o processo de matching irá tentar encontrar as correspondências com base nas regras definidas (por exemplo, FIFO).
}

function transactionEquals(t1: Transaction, t2: Transaction): boolean {
    
    return (
        t1.date.equals(t2.date) &&
        t1.type === t2.type &&
        (t1.asset && t2.asset
            ? t1.asset.equals(t2.asset)
            : t1.asset === t2.asset) &&
        t1.shares === t2.shares &&
        t1.amount === t2.amount &&
        t1.currency === t2.currency &&
        t1.broker.isEquals(t1.broker, t2.broker) &&
        (t1.exchangeRate || t2.exchangeRate ? t1.exchangeRate === t2.exchangeRate : true)
    );
}

type MatchedTransaction = {
    buy: Transaction;
    sell: Transaction;
    shares: number;
}



export { TransactionType, Transaction, MatchedTransaction, transactionEquals };